### PR TITLE
ROX-24374: Get Controls by RuleNames

### PR DESCRIPTION
**Pull request by SimonBaeumer**
## Description

Get Controls by RuleNames.
RuleNames are group by Name, Control and Standard to reduce the number of results. This improves performance later on to enrich API responses, otherwise the results would have been (Standard * Control * Rule Name * Cluster).

Specification:

1. Return all controls, standards and rule names for a list of rule names.
2. Only return distinct results, do not return results per cluster.

Follow-up PR: https://github.com/stackrox/stackrox/pull/11266/files

## Checklist
- [ ] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [ ] ~~Evaluated and added CHANGELOG entry if required~~
- [ ] ~~Determined and documented upgrade steps~~
- [ ] ~~Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~~

If any of these don't apply, please comment below.

## Testing Performed

 - Unit tests
 - More testing follows with the API implementations

### CODE REVIEW

File: central/complianceoperator/v2/rules/datastore/datastore_impl.go [hunk 1]
```diff
@@ -43,3 +47,33 @@ func (d *datastoreImpl) DeleteRulesByCluster(ctx context.Context, clusterID stri
...
+	builder.AddSelectFields(
+		search.NewQuerySelect(search.ComplianceOperatorControl),
+		search.NewQuerySelect(search.ComplianceOperatorStandard),
+		search.NewQuerySelect(search.ComplianceOperatorRuleName),
+	)
+
+	builder.AddExactMatches(search.ComplianceOperatorRuleName, ruleNames...)
+
+	// Add a group by clause to group the rule names by name, control and standard to reduce the result set.
+	builder.AddGroupBy(search.ComplianceOperatorRuleName, search.ComplianceOperatorControl, search.ComplianceOperatorStandard)
```
Comment (dashrews78) [inline central/complianceoperator/v2/rules/datastore/datastore_impl.go:70]: nit:
```suggestion
	builder.AddGroupBy(
	    search.ComplianceOperatorRuleName, 
	    search.ComplianceOperatorControl, 
	    search.ComplianceOperatorStandard
	)
```

File: pkg/search/options.go [hunk 2]
```diff
@@ -240,7 +240,10 @@ var (
 	ComplianceOperatorProfileRef             = newFieldLabel("Profile Ref ID")
 	ComplianceOperatorScanRef                = newFieldLabel("Scan Ref ID")
 	ComplianceOperatorRuleRef                = newFieldLabel("Rule Ref ID")
+	ComplianceOperatorRuleId                 = newFieldLabel("Compliance Rule ID")
```
Comment (dashrews78) [inline pkg/search/options.go:243]: Are these changes supposed to be on top of another PR?  Seems like the corresponding items in the storage proto are missing.

Comment (SimonBaeumer) [reply pkg/search/options.go:243]: This field is automatically added as a search field to the database. It does not need to be part of the rules storage proto.
Initially I had use it in my query but removed it to get distinct results. We'd like to keep it here that in the future someone can select the rule id which was not possible before.

File: central/complianceoperator/v2/rules/datastore/datastore_impl_test.go [hunk 3]
```diff
@@ -71,6 +71,99 @@ func (s *complianceRuleDataStoreTestSuite) TearDownTest() {
...
+			// setup test rules
+			for _, inputRule := range testCase.exampleRules {
+				err := s.dataStore.UpsertRule(ctx, inputRule)
+				s.Require().NoError(err)
+			}
+
+			controlQueryResults, err := s.dataStore.GetControlsByRuleNames(testCase.ctx, testCase.inputRuleNames)
+			s.Require().NoError(err)
+
+			for _, expectedRule := range s.getAllControls(testCase.exampleRules) {
```
Comment (lvalerom) [inline central/complianceoperator/v2/rules/datastore/datastore_impl_test.go:139]: [nit] this is a little bit confusing to read. We call `getAllControls` but the variable is called `expectedRule`. I would change to `getAllRuleControls` and `expectedRuleControl`.

File: central/complianceoperator/v2/rules/datastore/datastore_impl_test.go [hunk 4]
```diff
@@ -71,6 +71,99 @@ func (s *complianceRuleDataStoreTestSuite) TearDownTest() {
...
+			{Standard: "CIS-OCP", Control: "2.2.2"},
+			{Standard: "CIS-OCP", Control: "3.3.3"},
+			{Standard: "NERC-CIP", Control: "CIP-003-8 R5.1.1"},
+			{Standard: "NERC-CIP", Control: "CIP-555-9 R5.5.5"},
+		},
+	}
+
+	testCases := map[string]struct {
+		inputRuleNames []string
+		exampleRules   []*storage.ComplianceOperatorRuleV2
```
[the remark this answers is not in the document]
Comment (SimonBaeumer) [reply central/complianceoperator/v2/rules/datastore/datastore_impl_test.go:99]: expectedRule doesn't work here because the intention is to provide this as a fixture, or example data to test on. The expectation can currently be derived from the inputs, so we do not need to specify an additional expected field.
I've renamed the variable to `ruleFixtures` which  fits better.

File: pkg/search/options.go (hunk 2 above)
Comment (SimonBaeumer) [reply pkg/search/options.go]: Removed it.

File: central/complianceoperator/v2/rules/datastore/singleton.go [hunk 5]
```diff
@@ -14,7 +14,7 @@ var (
 )
 
 func initialize() {
-	dataStore = New(pgStore.New(globaldb.GetPostgres()))
+	dataStore = New(pgStore.New(globaldb.GetPostgres()), globaldb.GetPostgres())
```
Comment (dashrews78) [inline central/complianceoperator/v2/rules/datastore/singleton.go:17]: nit:

```suggestion
        pool := globaldb.GetPostgres()
	dataStore = New(pgStore.New(pool), pool)
```
